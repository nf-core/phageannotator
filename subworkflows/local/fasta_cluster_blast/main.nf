//
// Compare sequences by performing an all-v-all BLAST
//
include { GUNZIP                    } from '../../../modules/nf-core/gunzip/main'
include { BLAST_MAKEBLASTDB         } from '../../../modules/nf-core/blast/makeblastdb/main'
include { BLAST_BLASTN              } from '../../../modules/nf-core/blast/blastn/main'
include { ANICLUSTER_ANICALC        } from '../../../modules/local/anicluster/anicalc/main'
include { ANICLUSTER_ANICLUST       } from '../../../modules/local/anicluster/aniclust/main'
include { ANICLUSTER_EXTRACTREPS    } from '../../../modules/local/anicluster/extractreps/main'

workflow FASTA_CLUSTER_BLAST {

    take:
    fasta_gz    // [ [ meta ], fasta.gz ]   , assemblies/genomes (mandatory)
    min_ani     // val [ 0 - 100 ]          , minimum ANI for alignment to be counted
    min_qcov    // val [ 0 - 100 ]          , minimum query cover for clustering
    min_tcov    // val [ 0 - 100 ]          , minimum test cover for clustering

    main:
    ch_versions = Channel.empty()

    //
    // MODULE: Make BLASTN database
    //
    ch_blast_db = BLAST_MAKEBLASTDB ( fasta_gz ).db
    ch_versions = ch_versions.mix( BLAST_MAKEBLASTDB.out.versions )

    //
    // MODULE: Perform BLAST
    //
    ch_blast_txt = BLAST_BLASTN ( fasta_gz , ch_blast_db ).txt
    ch_versions = ch_versions = ch_versions.mix( BLAST_MAKEBLASTDB.out.versions )

    //
    // MODULE: Calculate average nucleotide identity (ANI) and alignment fraction (AF) based on BLAST
    //
    ch_ani_tsv = ANICLUSTER_ANICALC ( ch_blast_txt ).ani
    ch_versions = ch_versions.mix( ANICLUSTER_ANICALC.out.versions )

    // create input for ANICLUSTER_ANICALC
    ch_aniclust_input = fasta_gz.join( ch_ani_tsv )
        .multiMap { meta, fasta_gz, ani_tsv ->
            fasta: [ meta, fasta_gz ]
            ani: [ meta, ani_tsv ]
        }

    //
    // MODULE: Cluster virus sequences based on ANI and AF
    //
    ch_clusters_tsv = ANICLUSTER_ANICLUST (
        ch_aniclust_input.fasta,
        ch_aniclust_input.ani,
        min_ani,
        min_qcov,
        min_tcov
        ).clusters
    ch_versions = ch_versions.mix( ANICLUSTER_ANICLUST.out.versions )

    // create input for extracting cluster representatives
    ch_extractreps_input = fasta_gz.join( ch_clusters_tsv )

    //
    // MODULE: Extract cluster representatives
    //
    ch_anicluster_reps_fasta_gz = ANICLUSTER_EXTRACTREPS ( ch_extractreps_input ).representatives
    ch_versions = ch_versions.mix( ANICLUSTER_EXTRACTREPS.out.versions )

    emit:
    cluster_reps_fasta_gz   = ch_anicluster_reps_fasta_gz   // [ [ meta ], cluster_reps.fasta.gz ]  , FASTA file cluster representatives
    clusters_tsv            = ch_clusters_tsv               // [ [ meta ], clusters.tsv          ]  , TSV file containing cluster membership
    versions                = ch_versions                   // [ versions.yml ]

}
