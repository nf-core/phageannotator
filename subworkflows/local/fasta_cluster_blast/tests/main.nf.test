nextflow_workflow {

    name "Test Subworkflow: FASTA_CLUSTER_BLAST"
    script "subworkflows/local/fasta_cluster_blast/main.nf"
    workflow "FASTA_CLUSTER_BLAST"

    // Dependencies
    tag "BLAST_MAKEBLASTDB"
    tag "BLAST_BLASTN"
    tag "ANICLUSTER_ANICALC"
    tag "ANICLUSTER_ANICLUST"
    tag "ANICLUSTER_EXTRACTREPS"


    test("fasta.gz + 95 + 0 + 85") {

        when {
            params {
                anicluster_min_ani  = 95
                anicluster_min_qcov = 0
                anicluster_min_tcov = 85
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true)
                    ]
                )
                input[1] = params.anicluster_min_ani
                input[2] = params.anicluster_min_qcov
                input[3] = params.anicluster_min_tcov
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
