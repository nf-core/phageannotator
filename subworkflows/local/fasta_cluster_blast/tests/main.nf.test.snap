{
    "fasta.gz + 95 + 0 + 85": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_anicluster_representatives.fna.gz:md5,832eac6f8ccfb75cd524696eb1935141"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_clusters.tsv:md5,da359c01739cc9ddc70a435b7ddc1f3a"
                    ]
                ],
                "2": [
                    "versions.yml:md5,305338928ed290d9497d0a5d6f8bf3bb",
                    "versions.yml:md5,98963f24e789a741d265a3da5e6f1870",
                    "versions.yml:md5,eb2640269b37949f9a3fcadfca898450",
                    "versions.yml:md5,eb2640269b37949f9a3fcadfca898450",
                    "versions.yml:md5,f52ada5f485f0efe322e88949ac1bbe0"
                ],
                "cluster_reps_fasta_gz": [
                    [
                        {
                            "id": "test"
                        },
                        "test_anicluster_representatives.fna.gz:md5,832eac6f8ccfb75cd524696eb1935141"
                    ]
                ],
                "clusters_tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "test_clusters.tsv:md5,da359c01739cc9ddc70a435b7ddc1f3a"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,305338928ed290d9497d0a5d6f8bf3bb",
                    "versions.yml:md5,98963f24e789a741d265a3da5e6f1870",
                    "versions.yml:md5,eb2640269b37949f9a3fcadfca898450",
                    "versions.yml:md5,eb2640269b37949f9a3fcadfca898450",
                    "versions.yml:md5,f52ada5f485f0efe322e88949ac1bbe0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-22T14:53:55.502121639"
    }
}